### PR TITLE
bpo-19376: Added doc.

### PR DESCRIPTION
**Pull request by toanant**
```
[bpo-19376](https://bugs.python.org/issue19376): Added note mentioning `datetime.strptime()` without a year fails on Feb 29.
```

https://bugs.python.org/issue19376

### CODE REVIEW

**Comment by the-knights-who-say-ni** [comment]
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept your contribution by verifying you have signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).

Unfortunately our records indicate you have not signed the CLA. For legal reasons we need you to sign this before we can look at your contribution. Please follow [the steps outlined in the CPython devguide](https://devguide.python.org/pullrequest/#licensing) to rectify this issue.


You can [check yourself](https://check-python-cla.herokuapp.com/) to see if the CLA has been received.

Thanks again for your contribution, we look forward to reviewing it!

File: Doc/library/datetime.rst [hunk 1]
```diff
@@ -2252,6 +2252,13 @@ Notes:
    day of the week and the ISO year (``%G``) are specified in a
    :meth:`strptime` format string. Also note that ``%G`` and ``%Y`` are not
    interchangeable.
+   
+(9)
```
Comment (pganssle) [inline Doc/library/datetime.rst]: This footnote is not referenced anywhere. I think the behavior of `strptime` when no year, month or day components should be in the main body of the "Behavior of strftime() and strptime()" documentation.

**Review by pganssle — CHANGES_REQUESTED** [review]

**Comment by the-knights-who-say-ni** [comment]
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept your contribution by verifying you have signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).

Unfortunately we couldn't find an account corresponding to your GitHub username on [bugs.python.org](https://bugs.python.org/) (b.p.o) to verify you have signed the CLA (this might be simply due to a missing "GitHub Name" entry in your b.p.o account settings). This is necessary for legal reasons before we can look at your contribution. Please follow [the steps outlined in the CPython devguide](https://devguide.python.org/pullrequest/#licensing) to rectify this issue.


You can [check yourself](https://check-python-cla.herokuapp.com/) to see if the CLA has been received.

Thanks again for your contribution, we look forward to reviewing it!

File: Doc/library/datetime.rst [hunk 2]
```diff
@@ -2020,6 +2020,11 @@ equivalent to ``datetime(*(time.strptime(date_string, format)[0:6]))``, except
 when the format includes sub-second components or timezone offset information,
 which are supported in ``datetime.strptime`` but are discarded by ``time.strptime``.
 
+For :meth:`datetime.strptime` class method without year may fail for Feb 29,
```
Comment (taleinat) [inline Doc/library/datetime.rst]: Hi, thanks for working on this!

Please reword this in a shorter, simpler, clearer form. This one long sentence is difficult to understand.

Also, I'm not sure this should be so high this in this list, since it is a rather minor note about an obscure edge case. It should be at the end of this list, or near the end.

**Review by taleinat — CHANGES_REQUESTED** [review]

File: Doc/library/datetime.rst (hunk 2 above)
Comment (toanant) [reply Doc/library/datetime.rst]: Moved  near the end of section and reworded to shorter version.

**Review by taleinat — CHANGES_REQUESTED** [review]
Looking at this again, IMO we should add two notes:

1. Not proving a year leads to using 1900.
2. Feb. 29th without a year will fail due to the above.

These should be added at the end of the "Notes" section at the end of this page. It would probably be best to make this a single note with two short sentences.

I want you to know that seeing your suggested PR has helped me realize how I think this should be done, and hope you accept my requesting multiple revisions as a necessary part of the process.

**Comment by pganssle** [comment]
Also, @toanant I would like to echo @taleinat's thanks for your patience with this. I know it can feel like reviewers are being persnickety about minor things (particularly if they disagree with one another), and I hope that we have done a good job communicating to you that your contribution to the project is valuable.

**Comment by toanant** [comment]
@taleinat Added Notes as suggested.

**Comment by taleinat** [comment]
@toanant, I agree with @pablogsal: We should first mention that the default is `1900-01-01T00:00:00.000` and all non-supplied components will fall back to those values, and then as a followup mention that that means that parsing a string with Feb 29 but no year will fail.

**Comment by toanant** [comment]
> @toanant, I agree with @pablogsal: We should first mention that the default is `1900-01-01T00:00:00.000` and all non-supplied components will fall back to those values, and then as a followup mention that that means that parsing a string with Feb 29 but no year will fail.

Made changes accordingly with footnotes for Feb 29 example.

**Comment by csabella** [comment]
@taleinat and @pganssle Were there still outstanding questions on this?  It seems that @toanant addressed everything, but perhaps looking at this with fresh eyes will help get it merged.  :-)

Thanks!

**Comment by pganssle** [comment]
@csabella Thanks for the ping, sorry that I have let this fall off my radar. I think there are still some problems with the patch. I think a refactoring of the three paragraphs about this would be best, [per my earlier comments](https://github.com/python/cpython/pull/10243#issuecomment-435893997), but even going with the "add a paragraph" approach, there are some grammatical problems with the paragraph added. I will add inline comments.

File: Doc/library/datetime.rst [hunk 3]
```diff
@@ -2028,6 +2028,11 @@ For :class:`date` objects, the format codes for hours, minutes, seconds, and
 microseconds should not be used, as :class:`date` objects have no such
 values.  If they're used anyway, ``0`` is substituted for them.
 
+For :meth:`datetime.strptime` class method, default value is ``1900-01-01T00:00:00.000``
```
Comment (pganssle) [inline Doc/library/datetime.rst]:
```suggestion
For the :meth:`datetime.strptime` class method, the default value is ``1900-01-01T00:00:00.000``
```

File: Doc/library/datetime.rst [hunk 4]
```diff
@@ -2028,6 +2028,11 @@ For :class:`date` objects, the format codes for hours, minutes, seconds, and
...
+and when using :meth:`strptime`, any components requested in the format string but
+missing from the object itself (in the case of :meth:`datetime.date` and
+:meth:`datetime.time` objects) will be pulled from the default value.[#]_
```
Comment (pganssle) [inline Doc/library/datetime.rst]:
```suggestion
:meth:`datetime.time` objects) will be pulled from the default value. [#]_
```

A space before the reference is required to get it to render correctly.

File: Doc/library/datetime.rst [hunk 5]
```diff
@@ -2028,6 +2028,11 @@ For :class:`date` objects, the format codes for hours, minutes, seconds, and
 microseconds should not be used, as :class:`date` objects have no such
 values.  If they're used anyway, ``0`` is substituted for them.
 
+For :meth:`datetime.strptime` class method, default value is ``1900-01-01T00:00:00.000``
+and when using :meth:`strptime`, any components requested in the format string but
+missing from the object itself (in the case of :meth:`datetime.date` and
```
Comment (pganssle) [inline Doc/library/datetime.rst]: This part seems a bit confused. There is no `strptime` method on `date` and `time` objects (at least not yet). The preceding two paragraphs are about `strftime`, which *populates* a string from an object, and those objects may be missing the components specified in the format string.

In `strptime`, you are given a format string and a value string, and the object is populated from the values. The "default value" is talking about what happens when you are given a format string that is missing required components like year, month and day.

**Review by pganssle — CHANGES_REQUESTED** [review]
I still think some of this information can be consolidated between this paragraph and the two above it, but I don't think we need to make the perfect the enemy of the good here. Getting this additional paragraph into workable shape should be sufficient.

**Comment by toanant** [comment]
@pganssle  Thanks for your helpful inline comments, I've applied them accordingly.
Please let me know in case any modification needed for latest patch.

File: Doc/library/datetime.rst [hunk 6]
```diff
@@ -2028,6 +2028,10 @@ For :class:`date` objects, the format codes for hours, minutes, seconds, and
 microseconds should not be used, as :class:`date` objects have no such
 values.  If they're used anyway, ``0`` is substituted for them.
 
+For the :meth:`datetime.strptime` class method, the default value is ``1900-01-01T00:00:00.000``
+and when using :meth:`strptime`, any components requested in the format string but
+missing from the object itself will be pulled from the default value. [#]_
```
Comment (pganssle) [inline Doc/library/datetime.rst]: The "missing from the object itself" still doesn't make sense here, there's no object, what this is documenting is what happens when you do this:

```python
datetime.strptime("12:00:00", "%H:%M:%S")
```

Since the primary `datetime` constructor has required arguments, and `0` is not a valid value for year, month or day, we need to specify the default value for missing components.

So instead of "any components requested in the format string but missing from the object itself will be pulled from the default value", I think it should be "components not specified in the format string will be pulled from the default value."

The only reason I say "I think" is that people *may* get confused between the situation where neither the value string nor the format string contains the required components (which is what we are talking about) and the situation where the format string specifies a component but it's missing from the value string. Hopefully my suggested wording makes it clear that we're talking about the first case.

**Review by pganssle — APPROVED** [review]
Thanks for your patience @toanant, this looks good enough to me. Kind of hard to get this across clearly, so maybe there's a better way to do it, but I definitely think this change is a positive change.

**Comment by taleinat** [comment]
After further review, I suggest rewording the first paragraph to the following, which is more concise and hopefully clearer. If others approve, I will make this change and merge this PR.

For the `datetime.strptime` class method, the default value is `1900-01-01T00:00:00.000`: any components not specified in the format string will be pulled from the default value.

Suggested ReST:
```For the :meth:`datetime.strptime` class method, the default value is
``1900-01-01T00:00:00.000``: any components not specified in the format string
will be pulled from the default value. [#]_```

**Comment by toanant** [comment]
@pganssle @taleinat I've incorporated changes suggested by you. Do let me know any change needed for latest patch.

**Comment by csabella** [comment]
@pganssle and @taleinat, please review again after the last changes by @toanant.  Thanks!

**Review by pganssle — APPROVED** [review]
Looks good to me, sorry for the long delay in the review.

Thanks @csabella for the ping!

**Comment by csabella** [comment]
Thank you @toanant for the contribution and thank you @pganssle for the review and approval.  Also thanks to @taleinat for the reviews.  This has been merged!  🎉

**Comment by vstinner** [comment]
@csabella: It may be worth it to backport this doc change to 3.7, but @miss-islington looks sick :-(

**Comment by taleinat** [comment]
@vstinner, shouldn't this also be backported to 3.6 and 2.7?

**Comment by vstinner** [comment]
> @vstinner, shouldn't this also be backported to 3.6 and 2.7?

I became lazy and I stopped to backport doc **enhancements** to Python 2.7. Sorry Python 2.7 :-(

Python 3.6 no longer accept doc changes: https://devguide.python.org/#status-of-python-branches
